crypto/x509/x_pubkey.c

```diff
@@ -80,16 +83,34 @@ int X509_PUBKEY_set(X509_PUBKEY **x, EVP_PKEY *pkey)
             X509err(X509_F_X509_PUBKEY_SET, X509_R_METHOD_NOT_SUPPORTED);
             goto error;
         }
-    } else {
-        X509err(X509_F_X509_PUBKEY_SET, X509_R_UNSUPPORTED_ALGORITHM);
-        goto error;
+    } else if (pkey->pkeys[0].keymgmt != NULL) {
+        BIO *bmem = BIO_new(BIO_s_mem());
+        const char *serprop = "format=der,type=public";
+        OSSL_SERIALIZER_CTX *sctx =
+            OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(pkey, serprop);
+
+        if (OSSL_SERIALIZER_to_bio(sctx, bmem)) {
+            const unsigned char *der = NULL;
+            long derlen = BIO_get_mem_data(bmem, (char **)&der);
+
+            pk = d2i_X509_PUBKEY(NULL, &der, derlen);
+        }
+
+        OSSL_SERIALIZER_CTX_free(sctx);
+        BIO_free(bmem);
     }
 
+    if (pk == NULL)
+        goto unsupported;
+
     X509_PUBKEY_free(*x);
     *x = pk;
     pk->pkey = pkey;
     return EVP_PKEY_up_ref(pkey);
```

Comment:
This is pre-existing code but looking again it seems to me that the up-ref should happen before we amend `*x` so that if it fails we can free pk and leave the function with a failure code without any parameters being changed. As it stands if you get a failure from this function you don't know whether it is because the up-ref failed and therefore *x has been updated and needs to be freed, or whether some other error occurred.

Reply:
That's reasonable.  Will fix

Reply:
Done